azure_test.go

```diff
@@ -0,0 +1,176 @@
+
+	resp, err := http.DefaultClient.Do(req)
+	assert.NoError(t, err)
+	assert.Equal(t, http.StatusOK, resp.StatusCode)
+
+	var response VaultCredResponse
+	err = json.NewDecoder(resp.Body).Decode(&response)
+	assert.NoError(t, err)
+
+	clientID := response.Data.ClientId
+	clientSecret := response.Data.ClientSecret
+
+	// Introduce a delay between generating creds and using them
+	// time.Sleep(5 * time.Second)
+
+	newUUID := generateUUID()
+	var successes uint64
+	var wg sync.WaitGroup
+	creds, err := azidentity.NewClientSecretCredential(
+		tenantID, clientID, clientSecret, &azidentity.ClientSecretCredentialOptions{})
+	assert.NoError(t, err)
+
+	resourceGroupClient, err := armresources.NewResourceGroupsClient(subscriptionID, creds, nil)
+	assert.NoError(t, err)
+	n := 100
+	var authFailures uint64
+	wg.Add(n)
+	for i := 0; i < n; i++ {
+		go func() {
+			defer wg.Done()
+			count, err := createResourceGroup(ctx, resourceGroupClient, newUUID)
+			if err == nil {
+				atomic.AddUint64(&successes, 1)
+				_ = count
+			} else {
+				if strings.Contains(err.Error(), "unauthorized_client") {
+					atomic.AddUint64(&authFailures, 1)
+				} else {
+					fmt.Println(err)
+				}
+			}
+		}()
+	}
+	wg.Wait()
+	fmt.Println("Num failures", n-int(successes))
+	fmt.Println("Num Auth failures", authFailures)
+	assert.EqualValues(t, n, successes, successes)
+
+	err = cleanupResourceGroups(ctx, resourceGroupClient, newUUID)
+	assert.NoError(t, err)
+}
+
+func createResourceGroup(ctx context.Context, rgClient *armresources.ResourceGroupsClient, newUUID string) (armresources.ResourceGroupsClientCreateOrUpdateResponse, error) {
+	resp, err := rgClient.CreateOrUpdate(ctx, fmt.Sprintf("%v-%v", "vault-test", generateUUID()), armresources.ResourceGroup{
+		Location: to.StringPtr("West US"),
+		Tags:     map[string]*string{"created_by": to.StringPtr(fmt.Sprintf("vault-test-%s", newUUID))},
+	}, nil)
+	if err != nil {
+		return armresources.ResourceGroupsClientCreateOrUpdateResponse{}, err
+	}
+	return resp, nil
+}
+
+func cleanupResourceGroups(ctx context.Context, rgClient *armresources.ResourceGroupsClient, newUUID string) error {
+	pager := rgClient.NewListPager(&armresources.ResourceGroupsClientListOptions{
+		Filter: to.StringPtr(fmt.Sprintf("tagName eq 'created_by' and tagValue eq 'vault-test-%s'", newUUID)),
+		Top:    nil,
+	})
+	var counter uint64
+	for pager.More() {
+		page, err := pager.NextPage(ctx)
+		if err != nil {
+			return err
+		}
+		group, ctx := errgroup.WithContext(ctx)
+		for _, rg := range page.Value {
+			rg := rg
+			group.Go(func() error {
+				atomic.AddUint64(&counter, 1)
+				_, err := rgClient.BeginDelete(ctx, *rg.Name, nil)
+				var azerr *azcore.ResponseError
+				if errors.As(err, &azerr) {
+					if azerr.StatusCode == http.StatusNotFound {
+						return nil
+					}
+				}
+				return err
+			})
+		}
+		if err := group.Wait(); err != nil {
+			return err
+		}
+	}
+	fmt.Println("items marked for deletion", counter)
```

Comment:
Maybe we can use `t.Logf` instead of `fmt` here and everywhere else we call Println